spec/helpers/action-controller_spec.cr

```diff
@@ -0,0 +1,256 @@
+              readOnly: true
+            inner_schema:
+              $ref: '#/components/schemas/Model_InnerModel'
+            cast:
+              type: string
+              example: "1"
+        Model_InnerModel:
+          required:
+          - array_of_int
+          type: object
+          properties:
+            array_of_int:
+              type: array
+              items:
+                type: integer
+              writeOnly: true
+        Model_ComplexModel:
+          required:
+          - union_types
+          - free_form
+          - array_of_hash
+          type: object
+          properties:
+            union_types:
+              oneOf:
+              - type: object
+                additionalProperties:
+                  $ref: '#/components/schemas/Model_InnerModel'
+              - type: integer
+              - type: string
+            free_form:
+              type: object
+              additionalProperties: true
+            array_of_hash:
+              type: array
+              items:
+                type: object
+                additionalProperties:
+                  oneOf:
+                  - type: integer
+                  - type: string
+        Payload:
+          required:
+          - mandatory
+          - with_default
+          type: object
+          properties:
+            mandatory:
+              type: string
+            optional:
+              type: boolean
+            with_default:
+              type: string
+            with_default_nillable:
+              type: string
+      responses: {}
+      parameters: {}
+      examples: {}
+      requestBodies: {}
+      headers: {}
+      securitySchemes: {}
+      links: {}
+      callbacks: {}
+
+    YAML
+  end
+
+  it "should deserialise mandatory" do
+    res = HelloPayloadActionController.context(
+      method: "GET", route: "/hello",
+      route_params: {"mandatory" => "man"},
+      headers: {"Content-Type" => "application/json"}, &.create)
+
+    expected_body = Payload.new("man", nil, "default_value", "default_value_nillable")
+
+    res.status_code.should eq(200)
+    res.output.to_s.should eq(expected_body.to_json)
+  end
+
+  it "should set defaults" do
+    res = HelloPayloadActionController.context(
+      method: "GET", route: "/hello",
+      route_params: {
+        "mandatory"             => "man",
+        "optional"              => "true",
+        "with_default"          => "not_default",
+        "with_default_nillable" => "value",
+      },
+      headers: {"Content-Type" => "application/json"}, &.create)
+
+    expected_body = Payload.new("man", true, "not_default", "value")
+
+    res.status_code.should eq(200)
+    res.output.to_s.should eq(expected_body.to_json)
+  end
+
+  it "should raise if there is no mandatory param" do
+    expect_raises(KeyError) do
+      HelloPayloadActionController.context(method: "GET", route: "/hello", headers: {"Content-Type" => "application/json"}, &.create)
+    end
```

Comment:
Might be a good case for a more specific error

Reply:
updated